### PR TITLE
Persistent storage

### PR DESCRIPTION
**Pull request by shubhamdixit863**
Yes

Fixes https://github.com/numaproj-labs/numaplane/issues/88

### Modifications

Added a persistent volume and persistent volume claim to the GitSync controller to map the path with the cluster node path or host path. Additionally, modified the git clone process to utilize this persistent volume claim path for cloning repositories.

### Verification

The code has been tested in a local Kind cluster, where the cloned repository is visible in the cluster node's host path. Since the Kind cluster runs inside a Docker container, executing the `docker exec -it` command with shell access allows entry into the host path specified in the PV YAML file.

### CODE REVIEW

File: config/manager/persistent-volume-claim.yaml [hunk 1]
```diff
@@ -0,0 +1,11 @@
+apiVersion: v1
+kind: PersistentVolumeClaim
+metadata:
+  name: repo-pvc
+spec:
+  storageClassName: manual
+  accessModes:
+    - ReadWriteOnce
```
Comment (juliev0) [inline config/manager/persistent-volume-claim.yaml:8]: I wonder if we need `ReadWriteMany`? I see "ReadWriteOnce" means "the volume can be mounted as read-write by a single node. ReadWriteOnce access mode still can allow multiple pods to access the volume when the pods are running on the same node." We will have one Leader Controller Pod, and one Follower Controller Pod, and they can be on different nodes.

File: internal/git/util.go [hunk 2]
```diff
@@ -214,13 +214,11 @@ func isRootDir(path string) bool {
 // by default it will use /tmp as base directory
 // unless LOCAL_REPO_PATH env is set.
 func getLocalRepoPath(gitSync *v1alpha1.GitSync) string {
-	baseDir := os.Getenv("LOCAL_REPO_PATH")
+	// baseDir is persistent volume path on the cluster node
+	// TODO // should we get it from config variable ?
```
Comment (xdevxy) [inline internal/git/util.go:218]: yes, please add a config. Thanks!

File: config/manager/persistent-volume-claim.yaml (hunk 1 above)
Comment (xdevxy) [reply config/manager/persistent-volume-claim.yaml:8]: It seems fine as we should only read/write from the leader at a time?

**Review by xdevxy — COMMENTED** [review]
Have you verified in the local cluster the scenario that delete the leader pod and see if the new leader can continue the syncing without cloning again? Thanks!

File: config/manager/persistent-volume-claim.yaml (hunk 1 above)
Comment (juliev0) [reply config/manager/persistent-volume-claim.yaml:8]: I'm not sure? Is it?

**Comment by juliev0** [comment]
> Have you verified in the local cluster the scenario that delete the leader pod and see if the new leader can continue the syncing without cloning again? Thanks!

Do you mean regarding the `ReadWriteOnce` setting? But in order to truly test that, the two Pods would have to be on different Nodes, right?

File: config/manager/persistent-volume-claim.yaml [hunk 3]
```diff
@@ -0,0 +1,11 @@
+apiVersion: v1
+kind: PersistentVolumeClaim
+metadata:
+  name: repo-pvc
+spec:
+  storageClassName: manual
```
Comment (chandankumar4) [inline config/manager/persistent-volume-claim.yaml:6]: will there be already a storage class with name as `manual` ? @juliev0

File: config/manager/persistent-volume-claim.yaml [hunk 4]
```diff
@@ -0,0 +1,11 @@
+apiVersion: v1
+kind: PersistentVolumeClaim
+metadata:
+  name: repo-pvc
```
Comment (chandankumar4) [inline config/manager/persistent-volume-claim.yaml:4]: probably need to rename it something like: `numaplane-repo-pvc` or `numaplane-pvc`

File: config/manager/persistent-volume-claim.yaml (hunk 3 above)
Comment (chandankumar4) [reply config/manager/persistent-volume-claim.yaml:6]: Oh, I found an open issue related to it [here](https://github.com/kubernetes/website/issues/30912)

Comment (xdevxy) [reply config/manager/persistent-volume-claim.yaml:6]: https://github.com/kubernetes/website/issues/30912#issuecomment-1828679721

**Pull request closed without merging**